File: backend/src/main/java/com/hehe/thesocial/service/permission/PermissionService.java
package com.hehe.thesocial.service.permission;

import com.hehe.thesocial.dto.request.permission.PermissionRequest;
import com.hehe.thesocial.dto.response.permission.PermissionResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;

public interface PermissionService {
    Page<PermissionResponse> findAllPermissionsByPage(Pageable pageable);

    PermissionResponse findPermissionById(String id);

    @Transactional
    PermissionResponse createPermission(PermissionRequest permission);

    @Transactional
    PermissionResponse updatePermission(String id, PermissionRequest permission);

    @Transactional
    void deletePermission(String id);
}

File: admin-portal/src/services/admin/permissionService.ts
import apiClient from "@/config/axios";
import { ApiResponse, PagedResponse } from "@/types";

export interface PermissionResponse {
  id: string;
  name: string;
  permission?: string; // Some APIs use 'permission' instead of 'name'
  description: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface PermissionRequest {
  name?: string;
  permission?: string; // Some APIs use 'permission' instead of 'name'
  description: string;
}

class PermissionService {
  async getPermissions(
    page: number = 0,
    size: number = 12
  ): Promise<ApiResponse<PagedResponse<PermissionResponse>>> {
    const response = await apiClient.get(
      `/permissions?page=${page}&size=${size}`
    );
    return response.data;
  }

  async getPermissionById(
    id: string
  ): Promise<ApiResponse<PermissionResponse>> {
    const response = await apiClient.get(`/permissions/${id}`);
    return response.data;
  }

  async createPermission(
    request: PermissionRequest
  ): Promise<ApiResponse<PermissionResponse>> {
    const response = await apiClient.post("/permissions", request);
    return response.data;
  }

  async updatePermission(
    id: string,
    request: PermissionRequest
  ): Promise<ApiResponse<PermissionResponse>> {
    const response = await apiClient.put(`/permissions/${id}`, request);
    return response.data;
  }

  async deletePermission(id: string): Promise<ApiResponse<void>> {
    const response = await apiClient.delete(`/permissions/${id}`);
    return response.data;
  }
}

export const permissionService = new PermissionService();

